### PR TITLE
`geom_rect()` can derive corners from `x`/`width` or `y`/`height`

### LINKED ISSUE #5861: Feature request: flexible parametrisation of rectangles
Currently we have the following parametrisation for rectangles:

`geom_tile()` has `x`, `y`, `width` and `height`.
`geom_rect()` has `xmin`, `xmax`, `ymin` and `ymax`.

If you know two of the `x` or `y` related aesthetics, you can compute the rest regardless of which two are known. For example, if you know `xmin` and `width`, you can compute `xmax = xmin + width`.
Being able to abitrarily pick two aesthetics with the others imputed allows for more flexibility. For example the following two options for visualising the `presidential` dataset are both horrifying.

Using `geom_rect()` option, you're forced to manually resolve a discrete `y` into a continuous `y` and hack a continuous scale to look like a discrete scale:
``` r
library(ggplot2)

ggplot(presidential) +
  geom_rect(
    aes(xmin = start, 
        xmax = end, 
        ymin = match(name, unique(name)) - 0.4, 
        ymax = match(name, unique(name)) + 0.4)) +
  scale_y_continuous(
    breaks = seq_along(unique(presidential$name)),
    labels = unique(presidential$name)
  ) +
  scale_x_date()
```

Using the `geom_tile()` option you're forced to reparametrise the intervals, which can be a real pain for some classed objects like dates.

``` r
ggplot(presidential) +
  geom_tile(
    aes(x = structure((unclass(start) + unclass(end)) / 2, class = "Date"),
        y = name,
        width = end - start),
    height = 0.8
  ) +
  scale_y_discrete(
    limits = unique(presidential$name)
  )
#> Don't know how to automatically pick scale for object of type <difftime>.
#> Defaulting to continuous.
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

In the above cases, the ideal parameterisation is `xmin = start, xmax = end, y = name, height = 0.8`, but this currently isn't possible. Besides this, sometimes you might also wish to use mixed parameters and use, for example, `ymin` and `height` together.

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #5861.

Briefly, `geom_rect()` can derive the `xmin`/`xmax`/`ymin`/`ymax` parameters from `x`/`width` and `y`/`height` parameters.

The trickiest thing with this PR is what is understood as 'required aesthetics'. Currently, all the parameters mentioned above are 'optional aesthetics' and the corners ( `xmin`/`xmax`/`ymin`/`ymax`) are marked as 'non missing aesthetics'. The corners are subjected to a second `check_required_aesthetics()` to throw a warning.

A demo from the issue:

``` r
devtools::load_all("~/packages/ggplot2/")
#> ℹ Loading ggplot2

p <- ggplot(
  presidential,
  aes(xmin = start, xmax = end, y = forcats::fct(name))
)

p + geom_rect(height = 0.8)
```

However, if there is only 1 known aesthetic per position (in the example, just the `y` aesthetic), it'll throw an error due to missing aesthetics. It might be worth considering throwing a more tailored warning in this case.

``` r
p + geom_rect()
#> Error in `geom_rect()`:
#> ! Problem while setting up geom.
#> ℹ Error occurred in the 1st layer.
#> Caused by error in `setup_data()` at ggplot2/R/ggproto.R:190:17:
#> ! `geom_rect()` requires the following missing aesthetics: ymin and
#>   ymax.
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by teunbrand** [comment]
Possibly CI failing due to the R4.4.0 release? Will run again tomorrow.

**Comment by teunbrand** [comment]
I'm not really content with this PR as it seems like abusing the various aesthetics fields in the ggproto class. I'll put this as a draft for now.

**Comment by teunbrand** [comment]
TODO: Throw errors from setup_data

File: R/geom-rect.R [hunk 1]
```diff
@@ -31,7 +31,37 @@ GeomRect <- ggproto("GeomRect", Geom,
...
+        data[["y"]], data[["height"]] %||% params$height
+      )
+      i <- lengths(y) > 1
+      data[c("ymin", "ymax")[i]] <- y[i]
+    }
+    check_required_aesthetics(
+      self$non_missing_aes,
+      names(data),
+      snake_class(self)
+    )
```
Comment (teunbrand) [inline R/geom-rect.R:62]: Write out separate checks

**Comment by teunbrand** [comment]
I had to rewrite bits of `check_required_aesthetics()` to get around the hardcoded limit of 2 paired aesthetics, but I think this now working as it should without aesthetics field shenanigans.

**Review by thomasp85 — APPROVED** [review]
LGTM

**Pull request merged**